### PR TITLE
Support for OS("")

### PR DESCRIPTION
**Pull request by alexsaveliev**
This is how I see #3  :smile: 

- we do this so on Windows we can have a VFS which allows access to /C:/foo, /D:/foo, etc
- added test case meant to be run on Windows
- bonus: ability to run Travis CI builds on forked repository
- bonus: Appveyor integration

### CODE REVIEW

File: os_windows.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+var errInvalidPath = errors.New("Cannot resolve absolute path")
+
+func (root osFS) resolve(path string) (string, error) {
+	// Clean the path so that it cannot possibly begin with ../.
+	// If it did, the result of filepath.Join would be outside the
+	// tree rooted at root.  We probably won't ever see a path
+	// with .. in it, but be safe anyway.
+	path = strings.TrimPrefix(filepath.Clean(filepath.Join(string(root), path)), string(os.PathSeparator))
```
Comment (keegancsmith) [inline os_windows.go]: this code allows you to escape `root`. Say root is `C:/foo` and path is `../bar`. When this line run path will be `c:/bar`.

File: os_windows_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		slashfile      = "/" + tmpFile                       // absolute file path prefixed with slash, /C:/foo/bar
+		relfile        = strings.TrimPrefix(tmpFile, volume) // path w/o volume name, /foo/bar
+		shortfile      = "foo"                               // relative path, foo
+		slashshortfile = "/foo"                              // relative path prefixed with slash, /foo
+	)
+
+	// keys are mount points,
+	// values indicate if OS.Stat(path) should succeed for a given path
+	tests := map[string]map[string]bool{
+		empty: {
```
Comment (keegancsmith) [inline os_windows_test.go]: can you just write the strings here instead? I think the test case is easier to understand that way.

File: os_windows_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+				}
+				t.Errorf("stat `%s` on `%s` expected to %s", path, mount, expectation)
+				continue
+			}
+			if actual && !os.SameFile(fStat, stat) {
+				t.Errorf("stat `%s` on `%s` points to wrong file", path, mount)
+			}
+
+		}
+	}
```
Comment (keegancsmith) [inline os_windows_test.go:81]: I think you could make this a unit test by removing the tmp file + os operations and instead just testing the resolve function on windows. That would probably make this test much easier to understand since you can use simple hardcoded paths.

**Review by keegancsmith — CHANGES_REQUESTED** [review]
This change is great. It still feels a little strange to special case "", but I'm keen to go ahead with it. Something with less special edge cases is to instead create a new VFS called `ctxvfs.RootOS`. However, this works!

I have requested changes because the test case is hard to read as it is, so I haven't reviewed that closely to understand if we are getting the desired behaviour.

File: os_windows.go (hunk 1 above)
Comment (alexsaveliev) [reply os_windows.go]: fixed

File: os_windows_test.go (hunk 2 above)
Comment (alexsaveliev) [reply os_windows_test.go]: done

File: os_windows_test.go (hunk 3 above)
Comment (alexsaveliev) [reply os_windows_test.go:81]: done

File: os_windows_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+			"C:/folder/file":  "C:\\folder\\file",
+			"D:/folder/file":  "D:\\folder\\file",
+			"/C:/folder/file": "C:\\folder\\file",
+			"/folder/file":    "",
+			"file":            "",
+			"/file":           "",
+		},
+		"C:": {
+			"C:/folder/file":  "C:\\folder\\file",
+			"/C:/folder/file": "C:\\folder\\file",
```
Comment (keegancsmith) [inline os_windows_test.go:24]: why are you allowed to specify `C:/` and `/C:/`. Surely you should only be able to specify paths relative to `C:`? I expect the above to cases to resolve to the empty string.

Does the old `vfs.OS("C:")` have this behaviour? This crops up a few times in the test case, I'll just simply comment on them, but refer to this inline comment.

File: os_windows_test.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			"/folder/file":    "C:\\folder\\file",
+			"file":            "C:\\file",
+			"/file":           "C:\\file",
+			"D:/folder/file":  "",
+		},
+		"/C:": {
```
Comment (keegancsmith) [inline os_windows_test.go:30]: Does `vfs.OS` work with a root path like `/C:` on windows

File: os_windows_test.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+			"C:/folder/file":  "C:\\folder\\file",
+			"/C:/folder/file": "C:\\folder\\file",
```
Comment (keegancsmith) [inline os_windows_test.go:32]: Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

File: os_windows_test.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+			"/folder/file":    "C:\\folder\\file",
+			"file":            "C:\\file",
+			"/file":           "C:\\file",
+			"D:/folder/file":  "",
+		},
+		"C:/folder": {
+			"C:/folder/file":           "C:\\folder\\file",
+			"/C:/folder/file":          "C:\\folder\\file",
```
Comment (keegancsmith) [inline os_windows_test.go:40]: Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

File: os_windows_test.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+			"/folder/file":             "C:\\folder\\folder\\file",
+			"file":                     "C:\\folder\\file",
+			"/file":                    "C:\\folder\\file",
+			"C:/anoherfolder/file":     "",
+			"C:\\anoherfolder\\file":   "",
+			"/C:\\anoherfolder/file":   "",
+			"foo/bar/../../file":       "C:\\folder\\file",
+			"foo/bar\\\\\\..\\../file": "C:\\folder\\file",
+			"c:/Folder/fiLe":           "c:\\Folder\\fiLe",
```
Comment (keegancsmith) [inline os_windows_test.go:49]: Above line same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

File: os_windows_test.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		},
+		"/C:/folder": {
+			"C:/folder/file":         "C:\\folder\\file",
+			"/C:/folder/file":        "C:\\folder\\file",
```
Comment (keegancsmith) [inline os_windows_test.go:53]: Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

File: os_windows_test.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+			"/folder/file":           "C:\\folder\\folder\\file",
+			"file":                   "C:\\folder\\file",
+			"/file":                  "C:\\folder\\file",
+			"C:/anoherfolder/file":   "",
+			"C:\\anoherfolder/file":  "",
+			"/C:\\anoherfolder/file": "",
+		},
+		"C:/folder/": {
+			"C:/folder/file":         "C:\\folder\\file",
+			"/C:/folder/file":        "C:\\folder\\file",
```
Comment (keegancsmith) [inline os_windows_test.go:63]: Above two lines same as https://github.com/sourcegraph/ctxvfs/pull/4/files#r98679508

**Review by keegancsmith — CHANGES_REQUESTED** [review]

File: os_windows_test.go (hunk 4 above)
Comment (alexsaveliev) [reply os_windows_test.go:24]: > why are you allowed to specify C:/ and /C:/. Surely you should only be able to specify paths relative to C:? I expect the above to cases to resolve to the empty string.

I believe we should allow anything that located under the given mount point when combined with `root`. With the old implementation on Unix we can resolve "/foo/bar" to /"foo/bar" with mount point defined as a "/foo" so why do it different way on Windows?

>  Does the old vfs.OS("C:") have this behaviour? 

It doesn't. However on Unix old implementation gives the following results for `osFS.resolve` using `/foo` as a mount point:

"/foo/bar" =>  "/foo/bar", "bar =>  "/foo/bar"

I think that we should support both absolute and relative paths while they still located under a given mount point (or located system-wide in the cases of `/` or `""`)

File: os_windows_test.go (hunk 5 above)
Comment (alexsaveliev) [reply os_windows_test.go:30]: Nope, old `vfs.OS` doesn't work with `/C:`. With the old implementation:

```
got `\C:\foo` while resolving `foo` on `/C:`
got `\C:\foo` while resolving `/foo` on `/C:`
```
which cannot be translated to real OS path when trying to access file content

**Review by keegancsmith — APPROVED** [review]
